content.js

```diff
@@ -0,0 +1,78 @@
+/*
+ * Referer Modifier: Modify the Referer header in HTTP requests
+ * Copyright (C) 2017-2020 Fiona Klute
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+"use strict";
+
+// [CAVEAT]
+//  When the dynamic content script is run is undefined.
+//  engineConfig may be prefilled,
+//  or engineInstance.setConfig will be called later.
+var engineConfig;
+var engineInstance = new RefererModEngine(engineConfig);
```

Comment:
The problem is clear from the comment, but it should describe the (clever!) workaround, too. The `[CAVEAT]` on its own sounds to me like things might explode at any time. :wink: 

This also applies to the comment in the `registerContentScript()` function, but it's sufficient to have details in one place and a "see also" in the other.

Reply:
- Should I change `[CAVEAT]` to `[WORKAROUND]`?
- I personally think `engineConfig may be prefilled, or engineInstance.setConfig will be called later.` is an apt description here. Should I also include a link to Temporal Dead Zone (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let#temporal_dead_zone_tdz)? What do you have in mind for a description here?

Reply:
The comment should describe that we can't know which script (this file or the dynamically generated one) runs first, so the code has to handle both cases, and what the approach for that is. The reason for that code isn't obvious, so it needs to be explained, including why the value for the `document.referrer` has to be determined dynamically.

I don't really care about those labels, but if you want to leave one `[WORKAROUND]` would be much better.

Reply:
> The comment should describe that we can't know which script (this file or the dynamically generated one) runs first, so the code has to handle both cases, and what the approach for that is.

`When the dynamic content script is run is undefined.`
Or should it be something else? I don't know what better to write here.

> including why the value for the document.referrer has to be determined dynamically.

Because 1) there is no way to synchronously get settings from the background script and 2) `get referrer` can be called cross-window/document.
Where in the file should I put the explanation?
To me ...
Number 1 looks like it is obvious from the WebEnteision API documentation
Number 2 looks obvious from the DOM spec.

Do you have a suggestion as to what to write here?

Reply:
> `When the dynamic content script is run is undefined.`
> Or should it be something else? I don't know what better to write here.

The order relative to the static script being undefined is key here. So something like: "It is undefined whether this file will run before or after the dynamic content script (see registerContentScript() in background.js)."

As far as I'm concerned that's also why the referrer to return must be determined dynamically, if we could be sure the dynamic script with the config would run first I don't see why we couldn't use a static value.

> 2) `get referrer` can be called cross-window/document.

This is definitely non-obvious to me, and I'm still not sure about the possible impact here. Is that connected to the blank iframe question? Can you point me at the relevant documentation?

Reply:
> "It is undefined whether this file will run before or after the dynamic content script (see registerContentScript() in background.js)."

OK

> I don't see why we couldn't use a static value.

Wrong, do that cross-iframe code I mentioned in the "above" comment on a vanilla Firefox and see.

> This is definitely non-obvious to me, and I'm still not sure about the possible impact here. Is that connected to the blank iframe question? Can you point me at the relevant documentation?

Yes, it is connected.
Inspect `Reflect.getOwnPropertyDescriptor(Document.prototype)` on vanilla Firefox, and you will see that it is a getter, not a static value.

Reply:
> Can you point me at the relevant documentation?

https://searchfox.org/mozilla-central/source/dom/webidl/Document.webidl#128

Reply:
The comment should be just one paragraph, and two labels are confusing, best just remove them and combine the text so the connection between the sentences becomes clear:

"It is undefined whether static content scripts will run before or after dynamic content scripts. If the dynamic script ran first engineConfig is prefilled here, otherwise it will call engineInstance.setConfig later."